### ISSUE TITLE
ApplyT: Support convertible types

### ISSUE THREAD

**Issue by abhinav** [pulumi/pulumi#11784]
Given an `IDOutput`, `ApplyT` will not run a `func(string)` with it.

    bucketID := bucket.ID() // IDOutput
    out := bucketID.ApplyT(func(id string) pulumi.StringOutput {
        // ...
    })

The above panics because `ID` is not a `string`.
However, `ID` is safely convertible to `string`
because it's implemented as:

    type ID string

We should consider having ApplyT automatically do the conversion
for such cases.

Advantage: Makes usage of ApplyT easier.

Disadvantage:
Breaks the ApplyT abstraction a little.
ApplyT acts like a function call.
In Go, function calls do not automatically perform these conversions.
The following won't compile.

    var id ID = ...
    f := func(id string) string { ... }
    f(id)

**tl;dr: Should ApplyT automatically convert `f(id)` to `f(string(id))`?**

**Comment by justinvp** [comment]
If we did do this, would we special-case `ID` (and possibly `URN`? Any others?), or would we make it more general such that the conversion would always happen automatically if the function arg is safely convertible?

**Comment by abhinav** [comment]
I think we should _maybe_ special-case ID and URN for convenience.
It's possible to do the general thing (always convert if convertible) but it breaks the abstraction a little, and makes the expectations for ApplyT a bit fuzzy.
It's easier to digest the complexity of ApplyT if it doesn't have a bunch of asterisks next to it.

**Comment by Frassle** [comment]
Note we also have https://github.com/pulumi/pulumi/issues/11750 about the convertibility of IDs and ints, and if we did that it would probably raise if ID would also allow (int) methods for apply.

**Comment by AaronFriel** [comment]
I agree with `ApplyT` doing this if the output type is a newtype wrapper, but a generic `func Apply[T]` should not.

ApplyT already "breaks" monad laws (as many of us have discussed), as (using pseudo notation) using it on an (pseudo-notation) `Output[Output[T]]` will await internally to `T`.
